crates/aiken-lang/src/gen_uplc/builder.rs

```diff
@@ -1003,15 +1003,14 @@ pub fn unknown_data_to_type(term: Term<Name>, field_type: &Type) -> Term<Name> {
         Some(UplcType::Bool) => Term::snd_pair()
             .apply(Term::var("__pair__"))
             .delayed_choose_list(
-                Term::equals_integer()
-                    .apply(Term::integer(1.into()))
+                Term::less_than_equals_integer()
+                    .apply(Term::integer(2.into()))
                     .apply(Term::fst_pair().apply(Term::var("__pair__")))
                     .delayed_if_then_else(
-                        Term::bool(true),
+                        Term::Error,
                         Term::equals_integer()
-                            .apply(Term::integer(0.into()))
-                            .apply(Term::fst_pair().apply(Term::var("__pair__")))
-                            .delayed_if_then_else(Term::bool(false), Term::Error),
+                            .apply(Term::integer(1.into()))
+                            .apply(Term::fst_pair().apply(Term::var("__pair__"))),
```

Comment:
I am not sure to understand how this change makes it more correct? Now negative constructor values may count as `False` 🤔?

Reply:
Impossible due to ledger rules

Reply:
It has to come from a valid constr. We do unconstr_data

Reply:
Note that, nothing prevents doing something like:

```
let my_illed_constructor: Data = constr_data(-1, [])
``` 

This currently crashes the UPLC machine with:

```
   crates/uplc/src/machine/runtime.rs:762:43
       called `Result::unwrap()` on an `Err` value: TryFromBigIntError { original: () }
```

I wonder how the Haskell VM handles it. From the source code, I see no restrictions about the constructor being always positive. The CBOR decoder for Data though enforces that; but not the on-chain machine. So I can always think of weird edge-cases where people manually construct Data on-chain using builtins and shoot themselves in the foot.

Reply:
```
❯ uplc evaluate -i example.uplc
An error has occurred:
A builtin received a term argument when something else was expected
Caused by: mkCons -1
```

Reply:
It's not possible to serialize values as negative I don't see the reason to complicate codegen for something that is impossible to serialize

Reply:
Constr tags as negative in cbor